kong/pdk/tracing.lua

```diff
@@ -0,0 +1,452 @@
+
+  if attributes ~= nil then
+    if type(attributes) ~= "table" then
+      error("invalid attribute", 2)
+    end
+  end
+
+  if self.events == nil then
+    self.events = new_tab(4, 0)
+    self.events[0] = 0
+  end
+
+  local obj = new_tab(0, 3)
+  obj.name = name
+  obj.time_ns = time_ns or ffi_time_unix_nano()
+  if attributes then
+    obj.attributes = attributes
+  end
+
+  -- cache table length to avoid calling `luaH_getn`
+  local len = self.events[0] + 1
+  self.events[len] = obj
+  self.events[0] = len
+end
+
+-- Adds an error event to a Span
+--
+-- @module Span
+-- @tparam string err error string
+function span_mt:record_error(err)
+  if type(err) ~= "string" then
+    err = tostring(err)
+  end
+
+  self:add_event("exception", {
+    ["exception.message"] = err,
+  })
+end
+
+-- Adds an error event to a Span
+--
+-- @module Span
+-- @tparam number status status code, 0 unset, 1 ok, 2 error
+function span_mt:set_status(status)
+  if type(status) ~= "number" then
+    error("invalid status", 2)
+  end
+
+  self.status = status
+end
+
+
+--- Tracer
+-- @section tracer
+local tracer_mt = {}
+tracer_mt.__index = tracer_mt
+
+
+local tracer_cache = setmetatable({}, { __mode = "k" })
+
+local noop_tracer = {}
+noop_tracer.name = "noop"
+noop_tracer.start_span = function() return noop_span end
+noop_tracer.active_span = NOOP
+noop_tracer.set_active_span = NOOP
+noop_tracer.process_span = NOOP
+
+--- New Tracer
+local function new_tracer(name, options)
+  name = name or "default"
+
+  if tracer_cache[name] then
+    return tracer_cache[name]
+  end
+
+  local self = {
+    name = name, -- Instrumentation library name
+  }
+
+  options = options or {}
+  if options.noop then
+    return noop_tracer
+  end
+
+  options.sampling_rate = options.sampling_rate or 1.0
+  self.sampler = get_trace_id_based_sampler(options.sampling_rate)
+  self.active_span_key = name .. "_" .. "active_span"
+
+  --- Get the active span
+  -- Returns the root span by default
+  --
+  -- @function kong.tracing.new_span
+  -- @phases rewrite, access, header_filter, response, body_filter, log, admin_api
+  -- @treturn table span
+  function self.active_span()
+    if not base.get_request() then
+      return
+    end
+
+    return ngx.ctx[self.active_span_key]
+  end
+
+  --- Set the active span
+  --
+  -- @function kong.tracing.new_span
+  -- @phases rewrite, access, header_filter, response, body_filter, log, admin_api
+  -- @tparam table span
+  function self.set_active_span(span)
+    if not base.get_request() then
+      return
+    end
+
+    if span then
+      span.active = true
+    end
+
+    ngx.ctx[self.active_span_key] = span
+  end
+
+  --- Create a new Span
+  --
+  -- @function kong.tracing.new_span
+  -- @phases rewrite, access, header_filter, response, body_filter, log, admin_api
+  -- @tparam string name span name
+  -- @tparam table options TODO(mayo)
+  -- @treturn table span
+  function self.start_span(...)
+    if not base.get_request() then
```

Comment:
Basically, this can be called in any phases, for example, we have patched the DNS resolver, and the caller can use it in any phase, but we should not just throw an error and blocks the all the operations here, I think it better to create a NOOP span silently.
But I think it could be improved in the future.